File: PythonClient/ks/commands.py
# -*- coding: utf-8 -*-

# python imports
import sys
import struct
from enum import Enum

PY3 = sys.version_info > (3,)


class ECommandDirection(Enum):
	Up = 0
	Right = 1
	Down = 2
	Left = 3


class Move(object):

	@staticmethod
	def name():
		return 'Move'


	def __init__(self, id=None, direction=None):
		self.initialize(id, direction)
	

	def initialize(self, id=None, direction=None):
		self.id = id
		self.direction = direction
	

	def serialize(self):
		s = b''
		
		# serialize self.id
		s += b'\x00' if self.id is None else b'\x01'
		if self.id is not None:
			s += struct.pack('i', self.id)
		
		# serialize self.direction
		s += b'\x00' if self.direction is None else b'\x01'
		if self.direction is not None:
			s += struct.pack('b', self.direction.value)
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize self.id
		tmp0 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp0:
			self.id = struct.unpack('i', s[offset:offset + 4])[0]
			offset += 4
		else:
			self.id = None
		
		# deserialize self.direction
		tmp1 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp1:
			tmp2 = struct.unpack('b', s[offset:offset + 1])[0]
			offset += 1
			self.direction = ECommandDirection(tmp2)
		else:
			self.direction = None
		
		return offset


class PlantBomb(object):

	@staticmethod
	def name():
		return 'PlantBomb'


	def __init__(self, id=None, direction=None):
		self.initialize(id, direction)
	

	def initialize(self, id=None, direction=None):
		self.id = id
		self.direction = direction
	

	def serialize(self):
		s = b''
		
		# serialize self.id
		s += b'\x00' if self.id is None else b'\x01'
		if self.id is not None:
			s += struct.pack('i', self.id)
		
		# serialize self.direction
		s += b'\x00' if self.direction is None else b'\x01'
		if self.direction is not None:
			s += struct.pack('b', self.direction.value)
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize self.id
		tmp3 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp3:
			self.id = struct.unpack('i', s[offset:offset + 4])[0]
			offset += 4
		else:
			self.id = None
		
		# deserialize self.direction
		tmp4 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp4:
			tmp5 = struct.unpack('b', s[offset:offset + 1])[0]
			offset += 1
			self.direction = ECommandDirection(tmp5)
		else:
			self.direction = None
		
		return offset


class DefuseBomb(object):

	@staticmethod
	def name():
		return 'DefuseBomb'


	def __init__(self, id=None, direction=None):
		self.initialize(id, direction)
	

	def initialize(self, id=None, direction=None):
		self.id = id
		self.direction = direction
	

	def serialize(self):
		s = b''
		
		# serialize self.id
		s += b'\x00' if self.id is None else b'\x01'
		if self.id is not None:
			s += struct.pack('i', self.id)
		
		# serialize self.direction
		s += b'\x00' if self.direction is None else b'\x01'
		if self.direction is not None:
			s += struct.pack('b', self.direction.value)
		
		return s
	

	def deserialize(self, s, offset=0):
		# deserialize self.id
		tmp6 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp6:
			self.id = struct.unpack('i', s[offset:offset + 4])[0]
			offset += 4
		else:
			self.id = None
		
		# deserialize self.direction
		tmp7 = struct.unpack('B', s[offset:offset + 1])[0]
		offset += 1
		if tmp7:
			tmp8 = struct.unpack('b', s[offset:offset + 1])[0]
			offset += 1
			self.direction = ECommandDirection(tmp8)
		else:
			self.direction = None
		
		return offset

File: CSharpClient/Game/KS/Commands.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace KS.Commands
{
	public enum ECommandDirection
	{
		Up = 0,
		Right = 1,
		Down = 2,
		Left = 3,
	}
	
	public partial class Move : KSObject
	{
		public int? Id { get; set; }
		public ECommandDirection? Direction { get; set; }
		

		public Move()
		{
		}
		
		public new const string NameStatic = "Move";
		
		public override string Name() => "Move";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize Id
			s.Add((byte)((Id == null) ? 0 : 1));
			if (Id != null)
			{
				s.AddRange(BitConverter.GetBytes((int)Id));
			}
			
			// serialize Direction
			s.Add((byte)((Direction == null) ? 0 : 1));
			if (Direction != null)
			{
				s.Add((byte)((sbyte)Direction));
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize Id
			byte tmp0;
			tmp0 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp0 == 1)
			{
				Id = BitConverter.ToInt32(s, (int)offset);
				offset += sizeof(int);
			}
			else
				Id = null;
			
			// deserialize Direction
			byte tmp1;
			tmp1 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp1 == 1)
			{
				sbyte tmp2;
				tmp2 = (sbyte)s[(int)offset];
				offset += sizeof(sbyte);
				Direction = (ECommandDirection)tmp2;
			}
			else
				Direction = null;
			
			return offset;
		}
	}
	
	public partial class PlantBomb : KSObject
	{
		public int? Id { get; set; }
		public ECommandDirection? Direction { get; set; }
		

		public PlantBomb()
		{
		}
		
		public new const string NameStatic = "PlantBomb";
		
		public override string Name() => "PlantBomb";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize Id
			s.Add((byte)((Id == null) ? 0 : 1));
			if (Id != null)
			{
				s.AddRange(BitConverter.GetBytes((int)Id));
			}
			
			// serialize Direction
			s.Add((byte)((Direction == null) ? 0 : 1));
			if (Direction != null)
			{
				s.Add((byte)((sbyte)Direction));
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize Id
			byte tmp3;
			tmp3 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp3 == 1)
			{
				Id = BitConverter.ToInt32(s, (int)offset);
				offset += sizeof(int);
			}
			else
				Id = null;
			
			// deserialize Direction
			byte tmp4;
			tmp4 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp4 == 1)
			{
				sbyte tmp5;
				tmp5 = (sbyte)s[(int)offset];
				offset += sizeof(sbyte);
				Direction = (ECommandDirection)tmp5;
			}
			else
				Direction = null;
			
			return offset;
		}
	}
	
	public partial class DefuseBomb : KSObject
	{
		public int? Id { get; set; }
		public ECommandDirection? Direction { get; set; }
		

		public DefuseBomb()
		{
		}
		
		public new const string NameStatic = "DefuseBomb";
		
		public override string Name() => "DefuseBomb";
		
		public override byte[] Serialize()
		{
			List<byte> s = new List<byte>();
			
			// serialize Id
			s.Add((byte)((Id == null) ? 0 : 1));
			if (Id != null)
			{
				s.AddRange(BitConverter.GetBytes((int)Id));
			}
			
			// serialize Direction
			s.Add((byte)((Direction == null) ? 0 : 1));
			if (Direction != null)
			{
				s.Add((byte)((sbyte)Direction));
			}
			
			return s.ToArray();
		}
		
		public override uint Deserialize(byte[] s, uint offset = 0)
		{
			// deserialize Id
			byte tmp6;
			tmp6 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp6 == 1)
			{
				Id = BitConverter.ToInt32(s, (int)offset);
				offset += sizeof(int);
			}
			else
				Id = null;
			
			// deserialize Direction
			byte tmp7;
			tmp7 = (byte)s[(int)offset];
			offset += sizeof(byte);
			if (tmp7 == 1)
			{
				sbyte tmp8;
				tmp8 = (sbyte)s[(int)offset];
				offset += sizeof(sbyte);
				Direction = (ECommandDirection)tmp8;
			}
			else
				Direction = null;
			
			return offset;
		}
	}
} // namespace KS.Commands
